sound/soc/sof/intel/hda-dsp.c

```diff
@@ -336,7 +354,17 @@ int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
 	dev_vdbg(bus->dev, "D0I3C updated, register = 0x%x\n",
 		 snd_hdac_chip_readb(bus, VS_D0I3C));
 
-	return 0;
+	/* prevent power gating at D0 */
+	if (d0_substate == SOF_DSP_D0I0)
+		flags = SOF_PM_PPG | SOF_PM_PCG;
+
+	/* sending pm_gate IPC */
+	ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
+
+	return ret;
```

Comment:
I've mentioned this before, but this flow does not match what was discussed earlier.

In my notes and the arch specs, I see a sequence of 
a) send an IPC to tell the firmware to prepare to enter D0i3.
b) set the D0i3 bit.

Likewise for D0i3 exit, we are supposed to 
a) clear the D0i3 bit
b) send an IPC to tell the firmware to exit D0i3
 
We are not following this pattern...and during the D0i3 entry we pass exactly zero information to the firmware.

@lbetlej @mmaka1 am I missing something here?

Reply:
> I've mentioned this before, but this flow does not match what was discussed earlier.
> 
> In my notes and the arch specs, I see a sequence of
> a) send an IPC to tell the firmware to prepare to enter D0i3.
> b) set the D0i3 bit.

Here #a and #b should be swapped according to Lech's diagram here:
https://github.com/orgs/thesofproject/teams/sof-developers/discussions/6

> 
> Likewise for D0i3 exit, we are supposed to
> a) clear the D0i3 bit
> b) send an IPC to tell the firmware to exit D0i3

Here the sequence is correct, and actually aligned with my PR.

> 
> We are not following this pattern...and during the D0i3 entry we pass exactly zero information to the firmware.

We are still passing the flag in ext IPC register, with all bits set to 0, for d0i3 entry.

> 
> @lbetlej @mmaka1 am I missing something here?

Reply:
> > I've mentioned this before, but this flow does not match what was discussed earlier.
> > In my notes and the arch specs, I see a sequence of
> > a) send an IPC to tell the firmware to prepare to enter D0i3.
> > b) set the D0i3 bit.
> 
> Here #a and #b should be swapped according to Lech's diagram here:
> https://github.com/orgs/thesofproject/teams/sof-developers/discussions/6

@lbetlej can you double check please? This is not aligned with Rakesh's document, and the sequence also refers to D3 which is a separate flow.

> > Likewise for D0i3 exit, we are supposed to
> > a) clear the D0i3 bit
> > b) send an IPC to tell the firmware to exit D0i3
> 
> Here the sequence is correct, and actually aligned with my PR.
> 
> > We are not following this pattern...and during the D0i3 entry we pass exactly zero information to the firmware.
> 
> We are still passing the flag in ext IPC register, with all bits set to 0, for d0i3 entry.

That sounds like asking for trouble. an uninitialized value will cause D0i3 entry?

Reply:
> > We are still passing the flag in ext IPC register, with all bits set to 0, for d0i3 entry.
> 
> That sounds like asking for trouble. an uninitialized value will cause D0i3 entry?

We are sending power_gate IPC with:
CNL_DSP_REG_HIPCIDR = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE,
CNL_DSP_REG_HIPCIDD = SOF_PM_PCG (for D0)
                                     or    = 0 (for D0i3, we initialize the flag to be 0, it is not uninitialized:)
```
331+	u32 flags = 0;
```